marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/ApiSpec.hs

```diff
@@ -0,0 +1,247 @@
+{-# LANGUAGE OverloadedLists #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module Language.Marlowe.Runtime.Transaction.ApiSpec
+  ( spec
+  ) where
+
+import Language.Marlowe.Runtime.Transaction.Api
+  (NFTMetadata(..), NFTMetadataDetails(..), NFTMetadataFileDetails(..), mkMetadata, mkNFTMetadata)
+
+import Control.Arrow ((&&&), (***))
+import qualified Data.Aeson as Aeson
+import qualified Data.Aeson.Key as Aeson.Key
+import qualified Data.Aeson.KeyMap as Aeson.KeyMap
+import Data.Coerce (coerce)
+import Data.Foldable (find)
+import Data.Map (Map)
+import qualified Data.Map as Map
+import Data.Maybe (maybeToList)
+import qualified Data.Maybe as Maybe
+import Data.String (fromString)
+import Data.Text (Text)
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text.Encoding
+import qualified Data.Text.Encoding.Base16 as Base16
+import qualified Data.Vector as Vector
+import Language.Marlowe.Runtime.ChainSync.Api (PolicyId(PolicyId), TokenName(TokenName))
+import qualified Network.URI as Network (URI(..), URIAuth(..))
+import qualified Network.URI hiding (URI(..), URIAuth(..))
+import Test.Hspec (Spec, shouldBe, shouldSatisfy)
+import qualified Test.Hspec as Hspec
+import Test.Hspec.QuickCheck (prop)
+import Test.QuickCheck (Gen)
+import qualified Test.QuickCheck as Gen
+
+spec :: Spec
+spec = do
+  Hspec.describe "CIP-25 Metadata" do
+    prop "uriGen is valid" uriGenValidityTests
+    prop "NFTMetadataFileDetails is valid" cip25MetadataFileDetailsValidityTests
+    prop "NFTMetadataFileDetails is sound" cip25MetadataFileDetailsSoundnessTests
+    prop "NFTMetadataDetails is valid" cip25MetadataDetailsValidityTests
+    prop "NFTMetadata is valid" cip25MetadataValidityTests
+
+cip25MetadataFileDetailsSoundnessTests :: Gen.Property
+cip25MetadataFileDetailsSoundnessTests =
+  -- FIXME: Improve notNFTMetadataFileDetailsGen quality:
+  Gen.forAll notNFTMetadataFileDetailsGen \json -> do
```

Comment:
NB - usually when using manual generators it's also worth writing corresponding shrinkers and using `forAllShrink`